### ISSUE TITLE
Consider variance

### ISSUE THREAD

**Issue by striation** [stripe/bonsai#11]
As @johnynek notes, we could probably make the type classes contravariant.

https://github.com/stripe/bonsai/pull/10#discussion_r49503106

Given the specificity rules around implicit search, I'm not sure how useful contravariant type classes would be (for example, they didn't work out so well with `Ordering`) but it's worth a look at least.

**Comment by johnynek** [comment]
Actually I meant `IndexedBitSet[+A, +B]`, I think that will be safe if Array is private.

**Comment by tixxit** [comment]
Yeah, my general thing when I start is to keep everything invariant, since it's easier to add covariance than it is to take it away. But given how things have turned out, I don't see a good reason to not make them covariant.
